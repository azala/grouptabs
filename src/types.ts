export interface ActionMap {
  createOrUpdate: any[];
  delete: any[];
}

export enum DocumentType {
  TRANSACTION = 'transaction',
  INFO = 'info',
}

export enum TransactionType {
  SHARED = 'SHARED',
  DIRECT = 'DIRECT',
}

export interface Account {
  participant: string;
  amount: number;
}

export interface Transaction {
  id: string;
  type: DocumentType.TRANSACTION;
  tabId: string;
  timestamp: string;
  transactionType: TransactionType;
  description: string;
  date: string;
  participants: Account[];
}

export interface Info {
  id: string;
  type: DocumentType.INFO;
  tabId: string;
  name: string;
}

export type Entity = Transaction | Info;

export interface Tab {
  id: string;
  info: Info;
  mostRecentTransaction?: Transaction;
}
