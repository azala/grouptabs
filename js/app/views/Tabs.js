define([
	"./_Scene",
	"dojo/_base/declare",
	"dojo/_base/lang",
	"dojo/dom-class",
	"dijit/form/Button",
	"dijit/form/ValidationTextBox",
	"dojo/text!./templates/Tabs.html"
], function(_Scene, declare, lang, domClass, Button, ValidationTextBox, template){

return declare(_Scene, {

	templateString: template,

	name: "tabs",

	constructor: function(){
		this._tabButtons = []
	},

	postCreate: function(){
		this._createTabButtons()
	},

	refresh: function(){
		this._tabButtons.forEach(function(tabButton){
			tabButton.destroy()
		})
		this._tabButtons = []
		this._createTabButtons()
		if(this._getTabs().length){
			this.noTabsNode.style.display = 'none'
		}else{
			this.noTabsNode.style.display = ''
		}
	},

	_createTabButtons: function(){
		this._getTabs().forEach(function(tabName){
			var button = new Button({label: tabName})
			domClass.add(button.domNode, "full-width")
			button.onClick = lang.hitch(this, function(){
				this._onTabClick(tabName)
			})
			button.placeAt(this.tabsNode)
			this._tabButtons.push(button)
		}, this)
	},

	_getTabs: function(){
		var tabs = []
		this.app.store.query(function(){return true}).forEach(function(transaction){
			if(transaction.box && tabs.indexOf(transaction.box) === -1){
				tabs.push(transaction.box)
			}
		})
		return this._tabs = tabs
	},

	_setTab: function(tabName){
		this.app.setTab(tabName)
	},

	_onTabClick: function(tabName){
		this._setTab(tabName)
		this.close(this, "main")
	},

	_onOkClick: function(){
		var newTabName = this.newTabTextBox.get("value")
		this._setTab(newTabName)
		this.newTabTextBox.set("value", "")
		this.close(this, "main")
	}
})

})
