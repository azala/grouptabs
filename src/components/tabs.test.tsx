import React from "react";
import { render } from "@testing-library/react";
import Tabs from "./tabs";
import { DocumentType } from "../types";

it("renders tab view", () => {
  const { container } = render(
    <Tabs
      data={[
        {
          id: "123",
          info: {
            id: "123",
            type: DocumentType.INFO,
            tabId: "567",
            name: "TAB NAME",
          },
        },
      ]}
      onTabClick={jest.fn()}
      onCreateTabInputChange={jest.fn()}
      onCreateNewTab={jest.fn()}
      onImportTabInputChange={jest.fn()}
      onImportTab={jest.fn()}
    />
  );
  expect(container.firstChild).toMatchSnapshot();
});

it("renders empty tab view", () => {
  const { container } = render(
    <Tabs
      data={[]}
      onTabClick={jest.fn()}
      onCreateTabInputChange={jest.fn()}
      onCreateNewTab={jest.fn()}
      onImportTabInputChange={jest.fn()}
      onImportTab={jest.fn()}
    />
  );
  expect(container.firstChild).toMatchSnapshot();
});
