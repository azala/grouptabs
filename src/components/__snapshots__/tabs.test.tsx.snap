// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty tab view 1`] = `
<div
  class="scene scene-with-footer tabsScene hidden"
>
  <main>
    <div
      class="header"
    >
      <img
        alt=""
        id="logo"
        src="logo.png"
      />
      <h2>
        Grouptabs
      </h2>
    </div>
    <div
      class="empty-info"
    >
      <p>
        Track shared expenses in a group of people. Every group has its own tab like "Summer roadtrip" or "Badminton".
      </p>
      <p>
        Start by creating your first tab:
      </p>
    </div>
    <div
      class="row"
    >
      <form
        class="create-form"
      >
        <input
          class="full-width"
          placeholder="Tab name …"
          type="text"
          value=""
        />
        <button
          class="create"
        >
          Create
        </button>
      </form>
    </div>
    <div
      class="row"
    >
      <p
        class="fake-link"
      >
        Open shared tab
      </p>
    </div>
  </main>
  <footer>
    Version: 
    VERSION
     –
     
    <a
      href="https://grouptabs.net/"
      rel="noopener"
      target="_blank"
    >
      grouptabs.net
    </a>
  </footer>
</div>
`;

exports[`renders tab view 1`] = `
<div
  class="scene scene-with-footer tabsScene hidden"
>
  <main>
    <div
      class="header"
    >
      <img
        alt=""
        id="logo"
        src="logo.png"
      />
      <h2>
        Grouptabs
      </h2>
    </div>
    <div
      class="row tabs"
    >
      <button
        class="full-width"
      >
        TAB NAME
      </button>
    </div>
    <div
      class="row"
    >
      <form
        class="create-form"
      >
        <input
          class="full-width"
          placeholder="Tab name …"
          type="text"
          value=""
        />
        <button
          class="create"
        >
          Create
        </button>
      </form>
    </div>
    <div
      class="row"
    >
      <p
        class="fake-link"
      >
        Open shared tab
      </p>
    </div>
  </main>
  <footer>
    Version: 
    VERSION
     –
     
    <a
      href="https://grouptabs.net/"
      rel="noopener"
      target="_blank"
    >
      grouptabs.net
    </a>
  </footer>
</div>
`;
